Base64UrlEncoder performance improvements

cesarpayan opened the pull request:
Changed code paths to encode with bytes always. Temporaly breaking paths to EncodeString TBD if we can remove it.

Encode(byte, int, int) is optimized, as well as current UnsafeEncode (part of the EncodeString code path)

cesarpayan committed:
Changed code paths to encode with bytes always. Temporaly breaking paths to EncodeString TBD if we can remove it.

Encode(byte, int, int) is optimized, as well as current UnsafeEncode (part of the EncodeString code path)

brentschmaltz commented on src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs:
```diff
@@ -45,6 +45,18 @@ public static class Base64UrlEncoder
         private static char base64UrlCharacter62 = '-';
```
please change these statics to const.

dannybtsai commented on src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs:
```diff
@@ -92,7 +156,7 @@ public static string Encode(byte[] inArray)
             if (inArray == null)
                 throw LogHelper.LogArgumentNullException("inArray");
```
Use nameof(inArray) to avoid a mismatch between variable name and the name in the exception if the variable is renamed in the future.

dannybtsai commented on src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs:
```diff
@@ -76,7 +88,59 @@ public static string Encode(byte[] inArray, int offset, int length)
                 throw LogHelper.LogArgumentNullException("inArray");
 
-            return EncodeString(Convert.ToBase64String(inArray, offset, length));
+            int i;
```
nit: move i, j close to where they are used and remove the extra empty lines.

dannybtsai commented on src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs:
```diff
@@ -76,7 +88,59 @@ public static string Encode(byte[] inArray, int offset, int length)
             if (inArray == null)
```
Validate values (negative and 0) of offset and length.

cesarpayan committed:
Changing static variables to const

cesarpayan commented on src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs:
```diff
@@ -92,7 +156,7 @@ public static string Encode(byte[] inArray)
+            return Encode(inArray, 0, inArray.Length);
         }
 
         internal static string EncodeString(string str)
```
Should we preserve this method and its sub-methods or is it safe to remove them?  It is currently unreferenced by Wilson.

TimHannMSFT replied:
Seems if you were to check the projects listed as InternalsVisibleTo (\src\Microsoft.IdentityModel.Tokens\Properties\AssemblyInfo.cs) and none of them used it, removing would be a good thing imo.

TimHannMSFT commented on src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs:
```diff
@@ -76,7 +88,59 @@ public static string Encode(byte[] inArray, int offset, int length)
+            int i;
+
+            char[] table = s_base64Table;
+
```
nit: remove double blank line

brentschmaltz replied:
Even though it seems to be a hassle, we always add an error id to Token.LogMessages so that each error has an identifier that we can link to. In this case its not that important, but its a good habit.
We could even go so far as having different message IDs for different errors.
I have rarely seen a user complain about having very distinct error messages.

cesarpayan replied:
I wouldn't mind taking some time to add the LogMessages, is there an existing range I can make use of? or would I need to add a new one? In the case of the later, is there anything that needs to be documented?

-edit: Adding a generic ArgumentException message for now, if required I can add more specific message per case.

TimHannMSFT commented on src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs:
```diff
@@ -76,7 +88,59 @@ public static string Encode(byte[] inArray, int offset, int length)
             if (inArray == null)
                 throw LogHelper.LogArgumentNullException("inArray");
```
nit: I'm still not 100% on the team's style guide on this but Brent was mentioning these sorts of ArgNull checks are commonly performed with discards (_) and the null-coalescing operator (??), so:

_ = inArray ?? throw LogHelper.LogArgumentNullException("inArray");

TimHannMSFT commented on src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs:
```diff
@@ -76,7 +88,59 @@ public static string Encode(byte[] inArray, int offset, int length)
+            //Where we left off before
+            i = limit;
+
+            switch (lengthmod3)
```
nit: log an error on the default case, or Debug.Fail

cesarpayan replied:
The variable has 3 potential values (given that its value = length%3). In the case the value being 0, it wouldn't denote an error, but a case where no more operations are needed. Perhaps this would be more descriptive with if conditions instead of a switch?

TimHannMSFT replied:
Oops, my bad, just skimmed wrong. switch statement is fine, might help readability a tiny bit to just comment the default case but I think this is fine how you have it.

dannybtsai commented on src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs:
```diff
@@ -111,60 +175,46 @@ internal static string EncodeString(string str)
 #if !NET45
         private unsafe static string UnsafeEncode(string str)
```
Is "Encode" the right name? This function simply replace the 62 and 63 chars, so maybe rename it to UnsafeReplaceBase64UrlCharacters?

dannybtsai commented on src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs:
```diff
@@ -76,7 +88,59 @@ public static string Encode(byte[] inArray, int offset, int length)
+                byte d1 = inArray[i + 1];
+                byte d2 = inArray[i + 2];
+
+                output[j + 0] = table[(d0 & 0xfc) >> 2];
```
No need for the & op, can be optimized to:
table[d0 >> 2];

dannybtsai commented on src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs:
```diff
@@ -76,7 +88,59 @@ public static string Encode(byte[] inArray, int offset, int length)
+                byte d2 = inArray[i + 2];
+
+                output[j + 0] = table[(d0 & 0xfc) >> 2];
+                output[j + 1] = table[((d0 & 0x03) << 4) | ((d1 & 0xf0) >> 4)];
```
& is not needed, can be optimized to:
table[((d0 & 0x03) << 4) | (d1 >> 4)];

dannybtsai commented on src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs:
```diff
@@ -76,7 +88,59 @@ public static string Encode(byte[] inArray, int offset, int length)
+
+                output[j + 0] = table[(d0 & 0xfc) >> 2];
+                output[j + 1] = table[((d0 & 0x03) << 4) | ((d1 & 0xf0) >> 4)];
+                output[j + 2] = table[((d1 & 0x0f) << 2) | ((d2 & 0xc0) >> 6)];
```
Optimize to **table[((d1 & 0x0f) << 2) | d2 >> 6)];**

dannybtsai commented on src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs:
```diff
@@ -76,7 +88,59 @@ public static string Encode(byte[] inArray, int offset, int length)
+                        byte d0 = inArray[i];
+                        byte d1 = inArray[i + 1];
+
+                        output[j + 0] = table[(d0 & 0xfc) >> 2];
```
See comments above on optimization.

dannybtsai commented on src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs:
```diff
@@ -76,7 +88,59 @@ public static string Encode(byte[] inArray, int offset, int length)
+
+            int j = 0;
+
+            int lengthmod3 = length % 3;
```
Maybe add some comments to explain the logic of converting from 3 bytes to 4 bytes for anyone who is not familiar with the algorithm?

dannybtsai commented on src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs:
```diff
@@ -111,60 +175,46 @@ internal static string EncodeString(string str)
 #if !NET45
```
What is the reason for checking NET45?

cesarpayan committed:
Added resolutions to feedback from PR
Added edge error case check for UnsafeEncode
Added more parameter validations for Encode(byte[], int, int)

brentschmaltz commented on src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs:
```diff
@@ -73,10 +84,84 @@ public static string Encode(string arg)
-                throw LogHelper.LogArgumentNullException("inArray");
+            _ = inArray ?? throw LogHelper.LogArgumentNullException("inArray");
+
+            if (length < 0)
```
nit: for a single if clause, we sometimes skip the {.

brentschmaltz commented on src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs:
```diff
@@ -73,10 +84,84 @@ public static string Encode(string arg)
+
+            if (length < 0)
+            {
+                throw LogHelper.LogArgumentException<ArgumentException>(nameof(length), $"{nameof(length)} must be greater or equal than zero");
```
why does a method named 'LogArgumentException' need to be generic?
why the &lt;ArgumentException&gt;

cesarpayan replied:
Only reason I can think of, is for the usage of custom exceptions that could inherit from ArgumentException. But there's no case of it and only found the method being called in one occasion in a different place. I will move away from it to use LogHelper.LogExceptionMessage(new ArgumentException(..... which upon second inspection is the more widely way of logging argument exceptions. 

Perhaps it's worth bringing up the LogArgumentException method up for discussion.

brentschmaltz commented on src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs:
```diff
@@ -73,10 +84,84 @@ public static string Encode(string arg)
+
+            if (inArray.Length < offset)
+            {
+                throw LogHelper.LogArgumentException<ArgumentException>(nameof(offset), $"Invalid value for {nameof(offset)} parameter");
```
Its good practice to echo the value that was invalid.
So this message could read: "Invalid value for (nameof(offset)) parameter, was: '{offset}'."

Note we tend to put single quotes around such values as it helps identifying the value.

cesarpayan committed:
Adding custom generic message for argument exceptions
Code formatting

cesarpayan committed:
typo

pmaytak commented:
Are there any performance tests for comparison before and after with improvements?

cesarpayan committed:
Changing code path from EncodeString
Changed Argument to ArgumentOutOfRange Exception which would be more descriptive in this case

cesarpayan committed:
Removed unnecesary mask operations
use of nameof instead of strings denoting param name

cesarpayan reviewed (commented):
> Are there any performance tests for comparison before and after with improvements?
@pmaytak Yes there are. 
This is an example of using the both encoders (updated and current) using BenchmarkDotNet to measure their times of executions. In this case we are encoding a few strings and joining them together. 
![image](https://user-images.githubusercontent.com/86634096/132071248-977166cc-fb90-4631-b63d-3fbeeb64c1d2.png)

henrik-me commented:
@bgavrilMS

bgavrilMS commented:
Thanks for doing this work @cesarpayan. It looks like Encoding operations are now ~2x faster and take 50% less memory (in MSAL at least). Decode remains unchanged. Pretty good win. 

MSAL PR + perf tests including memory: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2879/
